src/tracelib/systrace.ts

```diff
@@ -0,0 +1,226 @@
+// codes is allowed, whereas using this blacklist when modifying source codes in
+// i.e. third-party libraries is forbidden. The list is usually used to shut up
+// chatty trace messages on third-party class methods.
+const PROPERTY_BLACKLIST = new Map(
+    // [['Foo', ['bar', 'xyz']],]
+);
+
+// Generates unique ids and cookies to identify tracing sessions or function
+// calls among execution contexts (main runtime, dedicated workers and etc).
+const idGenerator = new class {
+  private cookie: number = Date.now() & 0x7FFFFFFF;
+  private id: string = Math.random().toString(36).substr(2, 9);
+
+  /**
+   * An id is used to identify execution context.
+   *
+   * Relies on v8 pseudo-random number generator (PRNG).
+   * The random number is derived from an internal state, which is altered by
+   * a fixed algorithm for every new random number.
+   */
+  getUniqueId(): string {
+    return this.id;
+  }
+
+  /**
+   * A cookie is a rolling sequence being used to identify an unique
+   * asynchronous tracing session or function call.
+   */
+  getCookie(): number {
+    const cookie = this.cookie++;
+    this.cookie &= 0x7FFFFFFF;
+    return cookie;
+  }
+}();
+
+/**
+ * Method decorator for specifying that don't harness system trace
+ * to this method with a reason.
+ */
+export function DontTraceWithReason(reason: string = '') {
+  return function _(
+      target: object, property: string, descriptor: PropertyDescriptor) {
+    if (!target.hasOwnProperty(DONT_TRACE_PROPERTY)) {
+      Object.defineProperty(target, DONT_TRACE_PROPERTY, {value: [property]});
+    } else {
+      target[DONT_TRACE_PROPERTY].push(property);
+    }
+  };
+}
+
+/**
+ * Method decorator for specifying that don't harness system trace
+ * to this method.
+ */
+// tslint:disable-next-line: variable-name
+export const DontTrace = DontTraceWithReason();
+
+/**
+ * Class decorator for installing system tracing capability
+ * to a class and its subclasses.
+ */
+// tslint:disable-next-line:enforce-name-casing
+export function SystemTrace<T extends Constructor<{}>>(ctor: T) {
+  return class extends ctor {
+    constructor(...args) {
+      super(...args);
+      const clientClass: ReturnType<typeof getClientClass> = getClientClass();
+
+      // Don't harness system tracing when any of:
+      // a) clientClass is undefined, namely system tracing is not requested
+      //    via url parameter.
+      // b) re-entrance is detected.
+      if (clientClass &&
+          !this.constructor.hasOwnProperty(SYSTEM_TRACED_PROPERTY)) {
+        harnessSystemTracing(this, new clientClass());
+      }
+    }
+  };
+}
+
+/**
+ * Used at sources that cannot decorate class by @SystemTrace.
+ * Extends SystemTraceable to declare your class system-traceable,
+ * e.g. class Foo extends SystemTraceable {...}
+ */
+// tslint:disable-next-line: variable-name
+export const SystemTraceable = SystemTrace(class {});
+
+/**
+ * Used at sources that cannot decorate class by @SystemTrace.
```

Comment:
What is an example such source? I am just curious what catalyzed this.

Reply:
Hmm, 
in typescript, use @SystemTrace to decorate the traced classes.
in javascript, no decorators, so we have to do:
  class Foo extends SystemTrace(class {}) {...}
It's non-trivial to devs to declare their traced class like this, instead:
  class Foo extends SystemTraceable {...}